scib/metrics/morans_i.py

```diff
@@ -0,0 +1,149 @@
+    :param batch_key: Batch obs col in adata_pre.
+    :param n_hvg: How many top HVGs to use for Moran's I computation.
+    :param embed: Embedding to use for computation of neighbours in adata_post
+    :param rescale: Whether to rescale result to [0,1] so that 1 is better bio conservation
+    :param compare_pre: Whether to compute metric under scenario A instead of B.
+    :param hvgs: Custom list of genes to use for Moran's I computation (instead of hvgs
+        computed across batches).
+    :return: The resulting metric.
+    """
+    # Prepare adatas
+    # Get integrated connectivities for Moran's I
+    adata_post = adata_post.copy()
+    sc.pp.neighbors(adata_post, use_rep=embed)
+    # Prepare pre data
+    adata_pre = adata_pre.copy()
+    adata_pre.obs[batch_key] = adata_pre.obs[batch_key].astype("category")
+
+    # Get HVGs across samples
+    if hvgs is None:
+        hvgs = hvg_batch(
+            adata_pre, batch_key=batch_key, target_genes=n_hvg, flavor="cell_ranger"
+        )
+    else:
+        assert adata_pre.var_names.isin(hvgs).sum() == len(hvgs)
+
+    def compute_mi(data, hvgs):
+        """
+        Helper function for computing Moran's I on HVGS that are non-constant
+        and adding them to var
+        :param data: adata, result is added to var
+        :param hvgs: Genes for which to compute Moran's I
+        """
+        # Make sure that genes used for computation are non-constant
+        cond = (
+            np.array(np.mean(data[:, hvgs].X, axis=0))
+            != np.array(data[0, hvgs].X.todense())
+        ).squeeze()
+        hvgs_used = np.array(hvgs)[cond]
+        if len(hvgs) > len(hvgs_used):
+            print(
+                "Using subset (%i) of hvgs (%i) that are not constant in the data"
+                % (len(hvgs_used), len(hvgs))
+            )
+        # Compute Moran's I
+        morans_i = sc.metrics.morans_i(data[:, hvgs_used])
+        morans_i = pd.Series(morans_i, index=hvgs_used)
+        # This should not happen, just in case
+        if (morans_i > 1).any() or (morans_i < -1).any():
+            raise ValueError(
+                "Problems in computing Moran's I, value not between -1 and 1"
+            )
+        data.var["morans_i"] = morans_i
+
+    if compare_pre:
+        # Get per sample connectivities
+        adatas_sample = dict()
+        for sample in adata_pre.obs[batch_key].unique():
+            # Subset only to cells from 1 sample
+            adata_sample = adata_pre[adata_pre.obs[batch_key] == sample, :].copy()
+            # Compute HVG for each sample
+            sc.pp.highly_variable_genes(
+                adata_sample,
+                flavor="cell_ranger",
+                batch_key="study_sample",
+                n_top_genes=2000,
+            )
+            adata_sample_scl = adata_sample.copy()
+            # PP each sample: scaling, pca, neighbours
+            sc.pp.scale(adata_sample_scl, max_value=10)
```

Comment:
just out of curiosity: any reason for scaling here specifically or out of habit?

Reply:
Most of the code was adapted from Karin. @Hrovatin, do you have an opinion here?